polyglot/piranha/src/models/source_code_unit.rs

```diff
@@ -174,21 +167,74 @@ impl SourceCodeUnit {
   ///
   /// Note - Causes side effect. - Updates `self.ast` and `self.code`
   pub(crate) fn _apply_edit(
-    &mut self, range: Range, replacement_string: &str, parser: &mut Parser, handle_error: bool,
+    &mut self, range: Range, replacement_string: &str, parser: &mut Parser,
   ) -> InputEdit {
+    // Check if the edit is a `Delete` operation then delete trailing comma
+    let replace_range = if replacement_string.trim().is_empty() {
+      self.delete_trailing_comma(range)
+    } else {
+      range
+    };
     // Get the tree_sitter's input edit representation
     let (new_source_code, ts_edit) =
-      get_tree_sitter_edit(self.code.clone(), range, replacement_string);
+      get_tree_sitter_edit(self.code.clone(), replace_range, replacement_string);
     // Apply edit to the tree
     self.ast.edit(&ts_edit);
     self._replace_file_contents_and_re_parse(&new_source_code, parser, true);
-    // Handle errors, like removing extra comma.
-    if self.ast.root_node().has_error() && handle_error {
-      self.fix_syntax_for_comma_separated_expressions(parser);
+    if self.ast.root_node().has_error() {
+      let msg = format!(
+        "Produced syntactically incorrect source code {}",
+        self.code()
+      );
+      error!("{}", msg);
+      panic!("{}", msg);
     }
     ts_edit
   }
 
+  /// Deletes the trailing comma after the {deleted_range}
+  /// # Arguments
+  /// * `deleted_range` - the range of the deleted code
+  /// 
+  /// # Returns
+  /// code range of the closest node
+  ///
+  /// Algorithm: 
+  /// Get the node after the {deleted_range}'s end byte (heuristic 5 characters) 
+  /// Traverse this node and get the node closest to the range {deleted_range}'s end byte
+  /// IF this closest node is a comma, extend the {new_delete_range} to include the comma.
+  fn delete_trailing_comma(&mut self, deleted_range: Range) -> Range {
+    let mut new_deleted_range = deleted_range;
+
+    // Get the node immediately after the to-be-deleted code
+    if let Some(node_after_to_be_deleted_node) = self
+      .ast
+      .root_node()
+      .descendant_for_byte_range(deleted_range.end_byte, deleted_range.end_byte + 5)
```

Comment:
What does `+5` mean here? Is there a way to get the next token here, rather than trying to guess the byte indexes? What happens with:
```
STALE_FLAG              ,
```
? (GJF would get rid of that, but I think it's still valid Java)